rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,646 @@
+
+```
+# Whether to use TFRT-based implementation of the kernel C API.
+config_setting(
+    name = "tfrt_kernel_c_api",
+    define_values = {
+        "tfrt_kernel_c_api": "True",
+    },
+)
+```
+
+Most of the kernel C API implementation will be the same between the two with a
+few notable exceptions:
+
+*   TFRT Kernel Fallback implementation will cast `TF_OpKernelContext` and
+    `TF_OpKernelConstruction` to `TFRTOpKernelContext` and
+    `TFRTOpKernelConstruction` respectively.
+*   TFRT Kernel Fallback implementation will use Kernel Fallback registration
+    mechanism.
+
+### TFRT forwarding kernel registration using C API
+
+We plan to implement C API for TFRT kernel registration that calls TFRT Kernel
+Fallback registration mechanism. Note that this is analogous to TF Lite
+providing
+[their own C API registration mechanism](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/lite/c/common.h;l=739?q=tfliteregistration&ss=tensorflow%2Ftensorflow).
+
+```cpp
+TF_KernelBuilder* TF_NewKernelBuilder(
+    const char* op_name, const char* device_name,
+    void* (*create_func)(TF_OpKernelConstruction*),
+    void (*compute_func)(void*, TF_OpKernelContext*),
+    void (*delete_func)(void*)) {
+  TF_KernelBuilder* result = new TF_KernelBuilder;
+  result->create_function = create_func;
+  result->compute_function = compute_func;
+  result->delete_function = delete_func;
+  return result;
+}
+
+void TF_RegisterKernelBuilder(const char* name,
+                              TF_KernelBuilder* builder,
+                              TF_Status* status) {
+  auto* create_fn = builder->create_function;
+  auto* compute_fn = builder->compute_function;
+  auto* delete_fn = builder->delete_function;
+  auto create_kernel = [create_fn, compute_fn, delete_fn](
+      TFRTOpKernelConstruction* construction) {
+    return std::make_unique<tensorflow::TFRTCOpKernel>(
+        construction, create_fn, compute_fn, delete_fn);
+  };
+  ::tensorflow::TFRTKernelReg kernelinfo(create_kernel);
+  kernelinfo.type_constraints = builder->attr_to_type;
+  ::tensorflow::tfrt_kernel_factories->RegisterFactory(
+      name, kernelinfo);
+  tensorflow::TFRTOpRegisterer(tensorflow::TFRTOpMetaBuilder(name));
+  TF_DeleteKernelBuilder(builder);
+  TF_SetStatus(status, TF_OK, "");
+}
+```
+
+## TFRT integration
+
+Current preferred direction would generate a
+[BEF](https://github.com/tensorflow/runtime/blob/master/documents/binary_executable_format.md) file in advance and then run that
+file on a mobile device. Generated BEF file would have to call either native, TF
+Lite, runtime fallback or kernel fallback kernels and provide any glue logic in
+between (such as tensor conversions).
+
+We also need to consider how kernel or runtime fallback will be selected. This
+could be a parameter at BEF file creation step. It might also be good to package
+both runtime and kernel fallback implementations in a BEF file to be selected at
+runtime.
```

Comment:
I think this should be part of the TF stack that offline will look at the graph and registers to decide which kernel to use.
For example, long term we want to have mobile specific kernels defined (i.e. implemented and registered) via TF authoring APIs, and then we will prefer those to any pre-existing kernels that will be enabled by Kernel Fallback.

I'm not sure we would want to package kernels to be selected ad-hoc at runtime. Maybe I'm not understanding why you mean by it, or having a sample use-case would be useful.